### PR TITLE
Updated long infill method

### LINKED ISSUE #116: [API Change] Long and Wide Pandas Conversions
Current methods for converting pandas `DataFrame` between wide and long format do not use the index columns correctly. Instead, columns that would be the index are treated as any other column. It could be beneficial to convert these methods to use `Index` columns instead.

This change should result in less arguments needed for function calls, and slightly faster operation as more can be inferred (never mind the internal pandas speed gains). The following functions would be impacted:
* `caf.toolkit.pandas_utils.df_handling.`
  * `long_to_wide_infill()`
    * Less arguments needed, index / col will be inferred
  * `long_df_to_wide_ndarray()`
    * Less arguments needed, index / col will be inferred
  * `long_product_infill()`
    * Will return the df without resetting the index, leaving the columns in the index
  * `wide_to_long_infill()`
    * Will return the df without resetting the index, leaving the columns in the index

If going ahead with this update we need to be careful how the API change is fed back into users. Internal changes should easily be flagged by test failures, but other CAF packages may need updating to take these changes into account.

### PR DESCRIPTION
**Pull request by isaac-tfn**
Describe Changes
---
Fixes #116 

Task Checklist
----
- [x] Have unittests been added (testing individual functions and their edge cases)
- [x] Have integration tests been added (if applicable, to test modules of functionality)
- [ ] Updated RELEASE.md to reflect changes in PR
- [ ] Have new dependencies been added?
  - [ ] Have they been added to either `requirements.txt` or `requirements_dev.txt`.
  - [ ] Have they been added to `pyproject.toml`

### CODE REVIEW

**Comment by isaac-tfn** [comment]
Needs linting and checking. Tests pass but currently not compatible with python 3.9

**Comment by BenTaylor-TfN** [comment]
Can you edit the description and explain what you've changed and why please? It will make reviewing much easier if I know what I'm keeping an eye out for

**Comment by isaac-tfn** [comment]
> Can you edit the description in your opening comment and explain what you've changed and why please? It will make reviewing much easier if I know what I'm keeping an eye out for

I can't see an initial comment, will add to release.md?

File: src/caf/toolkit/translation.py [hunk 1]
```diff
@@ -575,27 +575,24 @@ def pandas_long_matrix_zone_translation(
     """
     # pylint: disable=too-many-arguments, too-many-locals
     # Init
+    if bool(index_col_2_out_name) != bool(index_col_1_out_name):
```
Comment (MattBuckley-TfN) [inline src/caf/toolkit/translation.py:578]: I don't think this causes any problems here but its worth noting some edge cases with None and an empty string both being False `bool(None) == bool("") == False`. Also, if integers are given that may also confuse things as `bool(0) == False` but any other number is True.

I think an explicit None check is usually better but in this case it would be quite verbose and I think the edge cases above are unlikely.

File: tests/test_translation.py [hunk 2]
```diff
@@ -933,135 +926,136 @@ def test_translation_correct(
...
-    ):
-        """Test that similar types are allowed in translation and data."""
-        pd_vector = request.getfixturevalue(pd_vector_str)
-        new_trans = pd_vector.translation.copy()
-        new_trans.from_col = new_trans.from_col.astype(np.int32)
-        result = translation.pandas_single_vector_zone_translation(
-            **(pd_vector.input_kwargs() | {"translation": new_trans.df})
-        )
-        pd.testing.assert_series_equal(result, pd_vector.expected_result)
+# class TestPandasVector:
```
Comment (MattBuckley-TfN) [inline tests/test_translation.py:929]: I think it is worth keeping a single test for the single vector translation even if its just a duplicate of one of the multi vector tests just incase the function gets edited in future.

As for the other tests in here can they just be removed instead of commented out.

**Review by MattBuckley-TfN — COMMENTED** [review]
Generally it looks good the functionality all seems to be working, I just have a few questions around what is expected in some edge cases.

My other comments are generally around making a few bits of the code simpler / easier to understand and clearing up docs and type annotation.

Depending on how much time you've got I think some of the suggestions could be moved to issues to look at in future, if required?

File: src/caf/toolkit/pandas_utils/df_handling.py [hunk 3]
```diff
@@ -446,122 +447,90 @@ def long_product_infill(
...
+        if len(mismatch) > 0:
+            raise ValueError(
+                f"{mismatch} levels were found in the input data, " f"but not in index_dict."
+            )
 
-    return df
+    full_ind = pd.MultiIndex.from_product(index_dict.values(), names=index_dict.keys())
+    if len(full_ind.names) == 1:
+        full_ind = full_ind.get_level_values(0)
+    filler = pd.DataFrame(data=["dummy"] * len(full_ind), index=full_ind, columns=["dummy"])
```
[the remark this answers is not in the document]
Comment (MattBuckley-TfN) [reply src/caf/toolkit/pandas_utils/df_handling.py:466]: Issue #141

[the remark this answers is not in the document]
Comment (isaac-tfn) [reply src/caf/toolkit/pandas_utils/df_handling.py:464]: OK, don't think I have time for now so will leave it but thanks

**Comment by isaac-tfn** [comment]
merged main into this one and now suddenly everything is failing. I think a lot is because of some stuff not available in 3.9, but some stuff may be due to more recent versions of e.g. mypy

**Comment by MattBuckley-TfN** [comment]
fixed linting issues, are we all happy to merge this? @BenTaylor-TfN @isaac-tfn

**Comment by BenTaylor-TfN** [comment]
Can you summarise what's actually changed please? Save me reading the many changes and possibly still being confused

**Comment by BenTaylor-TfN** [comment]
Happy with the changes here. Would just like to see a once over on the outstanding comments before accepting the merge.

File: src/caf/toolkit/pandas_utils/df_handling.py [hunk 4]
```diff
@@ -577,34 +546,38 @@ def long_to_wide_infill(
...
+            "for origin/production and another for destination/attraction."
+        )
+    if correct_cols is not None:
+        if correct_ind is not None:
+            ind_dict = {
+                matrix.index.names[0]: correct_ind,
+                matrix.index.names[1]: correct_cols,
+            }
+            matrix = long_product_infill(matrix, infill, index_dict=ind_dict)
+    unstacked = matrix.unstack(level=unstack_level, fill_value=infill)
```
[the remark this answers is not in the document]
Comment (MattBuckley-TfN) [reply src/caf/toolkit/pandas_utils/df_handling.py:584]: Just checked and long product infill doesn't allow providing one without the other so I'm raising an error in here if one is provided but not both

**Comment by MattBuckley-TfN** [comment]
Pylint is failing with the "Too many positional arguments" flag 17 times, is this something we want to fix or just up the limit to 10 (this will solve all but one of the issues).

The [pylint docs](https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/too-many-positional-arguments.html) suggest that we should force any arguments passed the first few as keyword-only, which does make sense to avoid confusion.

@isaac-tfn, @BenTaylor-TfN - what do you both think?

**Comment by BenTaylor-TfN** [comment]
What's changed in the code to make this appear now? 

As far as forcing keyword arguments, I'm happy for us to go for this if it makes sense. My general rule of thumb on whether something should not be forced as a keyword is "Will this argument need to be passed in for 80% of the calls the this function?". If not, it can probably be a keyword one.

**Comment by MattBuckley-TfN** [comment]
> What's changed in the code to make this appear now?
> 
> As far as forcing keyword arguments, I'm happy for us to go for this if it makes sense. My general rule of thumb on whether something should not be forced as a keyword is "Will this argument need to be passed in for 80% of the calls the this function?". If not, it can probably be a keyword one.

Looks like nothing, it worked for commit [436b967](https://github.com/Transport-for-the-North/caf.toolkit/pull/117/commits/436b967018badb056bc05e53bc1bf2f422713837) with pylint v3.2.7 but doesn't work on v3.3.1. Running pylint=3.27 on the current repo doesn't show the issues.

**Comment by MattBuckley-TfN** [comment]
One caveat with forcing keyword arguments is that its technically a public API change so it might break stuff elsewhere

**Comment by BenTaylor-TfN** [comment]
> > What's changed in the code to make this appear now?
> > As far as forcing keyword arguments, I'm happy for us to go for this if it makes sense. My general rule of thumb on whether something should not be forced as a keyword is "Will this argument need to be passed in for 80% of the calls the this function?". If not, it can probably be a keyword one.
> 
> Looks like nothing, it worked for commit [436b967](https://github.com/Transport-for-the-North/caf.toolkit/pull/117/commits/436b967018badb056bc05e53bc1bf2f422713837) with pylint v3.2.7 but doesn't work on v3.3.1. Running pylint=3.27 on the current repo doesn't show the issues.

Looks like it's an update with the PyLint checks then in reality?

**Comment by MattBuckley-TfN** [comment]
Yes, something they must have changed between v3.2 and v3.3. I'll limit pylint in requirements to 3.2
